lib/galaxy/tools/parameters/__init__.py

```diff
@@ -312,7 +312,82 @@ def populate_state(request_context, inputs, incoming, state, errors={}, prefix='
+
+        elif input.type == 'conditional':
+            # if input.value_ref and not input.value_ref_in_group:
+            #     test_param_key = prefix + input.test_param.name
+            # else:
+            #     test_param_key = group_prefix + input.test_param.name
+            test_param_value = incoming.get(input.name).get(input.test_param.name)
+            value, error = check_param(request_context, input.test_param, test_param_value, context, simple_errors=simple_errors) if check else [test_param_value, None]
+            if error:
+                errors[input.test_param.name] = error
+            else:
+                try:
+                    current_case = input.get_current_case(value)
+                    group_state = state[input.name] = {}
+                    populate_state(request_context, input.cases[current_case].inputs, incoming.get(input.name), group_state, errors, context=context, check=check, simple_errors=simple_errors, legacy_param_form=False)
+                    group_state['__current_case__'] = current_case
+                except Exception:
+                    errors[input.test_param.name] = 'The selected case is unavailable/invalid.'
+            group_state[input.test_param.name] = value
+
+        elif input.type == 'section':
+            populate_state(request_context, input.inputs, incoming.get(input.name), group_state, errors, context=context, check=check, simple_errors=simple_errors)
+
+        elif input.type == 'upload_dataset':
+            file_count = input.get_file_count(request_context, context)
+            while len(group_state) > file_count:
+                del group_state[-1]
+            while file_count > len(group_state):
+                new_state = {'__index__' : len(group_state)}
+                for upload_item in input.inputs.values():
+                    new_state[upload_item.name] = upload_item.get_initial_value(request_context, context)
+                group_state.append(new_state)
+            for i, rep_state in enumerate(group_state):
+                rep_index = rep_state['__index__']
+                rep_prefix = '%s_%d|' % (input.name, rep_index)
+                populate_state(request_context, input.inputs, incoming.get(rep_prefix, []), rep_state, errors, context=context, check=check, simple_errors=simple_errors)
+        else:
+            param_value = _get_incoming_value(incoming, input.name, state.get(input.name))
+            value, error = check_param(request_context, input, param_value, context, simple_errors=simple_errors) if check else [param_value, None]
+            if error:
+                errors[input.name] = error
+            state[input.name] = value
+
+
+def _populate_state_legacy(request_context, inputs, incoming, state, errors={}, prefix='', context=None, check=True, simple_errors=True):
```

Comment:
The recursive calls below should call directly `_populate_state_legacy()` instead of `_populate_state()` .